vortex-array/src/arrow/dtype.rs

```diff
@@ -91,3 +104,192 @@ impl FromArrowType<&Field> for DType {
         }
     }
 }
+
+/// Convert a Vortex [struct DType][DType] to an Arrow [Schema].
+pub fn infer_schema(dtype: &DType) -> VortexResult<Schema> {
+    let DType::Struct(struct_dtype, nullable) = dtype else {
+        vortex_bail!("only DType::Struct can be converted to arrow schema");
+    };
+
+    if *nullable != Nullability::NonNullable {
+        vortex_bail!("top-level struct in Schema must be NonNullable");
+    }
+
+    let mut builder = SchemaBuilder::with_capacity(struct_dtype.names().len());
+    for (field_name, field_dtype) in struct_dtype
+        .names()
+        .iter()
+        .zip(struct_dtype.dtypes().iter())
+    {
+        builder.push(FieldRef::from(Field::new(
+            field_name.to_string(),
+            infer_data_type(field_dtype)?,
+            field_dtype.is_nullable(),
+        )));
+    }
+
+    Ok(builder.finish())
+}
+
+pub fn infer_data_type(dtype: &DType) -> VortexResult<DataType> {
+    Ok(match dtype {
+        DType::Null => DataType::Null,
+        DType::Bool(_) => DataType::Boolean,
+        DType::Primitive(ptype, _) => match ptype {
+            PType::U8 => DataType::UInt8,
+            PType::U16 => DataType::UInt16,
+            PType::U32 => DataType::UInt32,
+            PType::U64 => DataType::UInt64,
+            PType::I8 => DataType::Int8,
+            PType::I16 => DataType::Int16,
+            PType::I32 => DataType::Int32,
+            PType::I64 => DataType::Int64,
+            PType::F16 => DataType::Float16,
+            PType::F32 => DataType::Float32,
+            PType::F64 => DataType::Float64,
+        },
+        DType::Utf8(_) => DataType::Utf8,
+        DType::Binary(_) => DataType::Binary,
+        DType::Struct(struct_dtype, _) => {
+            let mut fields = Vec::with_capacity(struct_dtype.names().len());
+            for (field_name, field_dt) in struct_dtype
+                .names()
+                .iter()
+                .zip(struct_dtype.dtypes().iter())
+            {
+                fields.push(FieldRef::from(Field::new(
+                    field_name.to_string(),
+                    infer_data_type(field_dt)?,
+                    field_dt.is_nullable(),
+                )));
+            }
+
+            DataType::Struct(Fields::from(fields))
+        }
+        // There are four kinds of lists: List (32-bit offsets), Large List (64-bit), List View
+        // (32-bit), Large List View (64-bit). We cannot both guarantee zero-copy and commit to an
+        // Arrow dtype because we do not how large our offsets are.
+        DType::List(..) => vortex_bail!("Unsupported dtype: {}", dtype),
```

Comment:
For posterity - we will canonicalize to List View. We should figure out what to do about fixed size lists though